Enforce pan/tilt limits, and related improvements.

kmdalal opened the pull request:
## Summary

- **IMPORTANT**: This branch must be used with the same branch in `ow_simulator`. 
- PLEXIL now enforces pan range of [-3.2, 3.2] and tilt range of (-pi/2, pi/2).
- Angles are compared with regard for equivalences.
- Tolerances now match those used in simulator.
- Improved pan/tilt test plan.

## Test

1. Start any simulator world.
2. Adjust Gazebo to see the antenna clearly.
3. Run the plan TestAntennaMovement.   E.g.
```
$ roslaunch ow_plexil ow_exec.launch plan:=TestAntennaMovement.plx
```
All 13 tests should succeed.  They command a wide range of pans and tilts.

AstroStucky commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -308,6 +303,18 @@ void OwInterface::antennaFaultCallback
   updateFaultStatus (msg->value, m_panTiltErrors, "ANTENNA", "AntennaFault");
 }
 
+static double normalize_degrees (double angle)
+{
+  static double pi_degrees = R2D * M_PI;
+  while (angle > pi_degrees) {
+    angle -= 2.0 * pi_degrees;
+  }
+  while (angle < -pi_degrees) {
+    angle += 2.0 * pi_degrees;
+  }
+  return angle;
+}
```
[optional] This could be done in the same form as in utils.py with `fmod`.

kmdalal replied:
Agreed, done.

AstroStucky commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -336,12 +343,12 @@ void OwInterface::jointStatesCallback
       if (joint == Joint::antenna_pan) {
-        m_currentPan = position * R2D;
+        m_currentPan = normalize_degrees (position * R2D);
         managePanTilt (Op_PanAntenna, m_currentPan, m_goalPan, m_panStart);
         publish ("PanDegrees", m_currentPan);
      }
       else if (joint == Joint::antenna_tilt) {
-        m_currentTilt = position * R2D;
+        m_currentTilt = normalize_degrees (position * R2D);
```
Should this be un-normalized like the tracked pan and tilt are now in the antennat pan tilt action server?

kmdalal replied:
It seems so, and it works, done.

AstroStucky commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -401,6 +388,14 @@ void OwInterface::managePanTilt (const string& opname,
 
 ///////////////////////// Antenna/Camera Support ///////////////////////////////
+
+bool OwInterface::anglesEquivalent (double deg1, double deg2, double tolerance)
+{
+  return (fabs(deg1 - deg2) <= tolerance ||
+          fabs((deg1 - 360) - deg2) <= tolerance ||
+          fabs((deg1 + 360) - deg2) <= tolerance);
+}
```
Would work better as the form discussed in the other PR.

kmdalal replied:
Agreed, done.

AstroStucky reviewed (changes requested):
Test worked fine. A few code comments, which are essentially repeats of the ones I made in the ow_simulator sibling PR.

AstroStucky commented:
@kmdalal I think you forgot to push. Once you do I'll re-review.

kmdalal commented:
> @kmdalal I think you forgot to push. Once you do I'll re-review.

@AstroStucky you were right, and for some reason I just saw this message now.  Just pushed.

AstroStucky commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -32,18 +32,18 @@ using std::shared_ptr;
 
 // C
-#include <cmath>  // for M_PI and fabs
+#include <math.h>  // for M_PI, fabs, fmod
```
Why switch to `math.h`? The preferred library to use here would be `cmath`.

kmdalal replied:
I read just the opposite in an upvoted StackOverflow article but don't want to argue such a small point -- changed it back to cmath.

AstroStucky replied:
Interesting... I have only heard arguments for using only cmath in C++ code. What I have read says that the c* headers are like of the original C libraries adapted for C++ code, and their *.h counterparts should be considered deprecated. Granted, I think the only way using *.h over c* could cause trouble is if people are not careful with their type conversions, in particular when using intrinsic types that did not exist in C (e.g. https://cplusplus.com/forum/general/5207/).

kmdalal replied:
The argument I read is that they are equivalent, except that math.h guarantees definition in the global namespace (with possible definition in std), while cmath guarantees definition in std (with possible definition in global).  I wanted to use the functions without a namespace prefix, so math.h seemed more portable for this choice.  But portability is not much of an issue for OceanWATERS, and this issue is too small to ponder too much :)

AstroStucky commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -391,9 +387,7 @@ void OwInterface::managePanTilt (const string& opname,
-  return (fabs(deg1 - deg2) <= tolerance ||
-          fabs((deg1 - 360) - deg2) <= tolerance ||
-          fabs((deg1 + 360) - deg2) <= tolerance);
+  return normalize_degrees(fabs(deg1 - deg2)) <= tolerance;
```
Possible bug since `normalize_degrees` is capable of being passed a position number and returning a negative number. This becomes more rigorous if we compute the absolute value after normalization. This idiom was my recommendation, so I take full responsibility :laughing:
```suggestion
  return fabs(normalize_degrees(deg1 - deg2)) <= tolerance;
```

AstroStucky reviewed (changes requested):
Some comments, one required change because I worry it could a bug. Re-tested and it passed.

kmdalal commented:
I augmented this work by tightening the tilt limits as decided in Dev Day this week, and changed the test plan to use these limits.  I also added the same pan/tilt input tolerances used in the simulator's action client scripts.   The same test given in the description can be repeated to verify these changes.

AstroStucky commented:
The test plan for this did not succeed. Perhaps changes need to be carried over to that plan?
```
[ INFO] [1667849267.986605175, 1916524893.224999905]: Starting PLEXIL executive node...
[ INFO] [1667849277.993170744, 1916524903.224999905]: Executive node started, ready for PLEXIL plans.
[ INFO] [1667849277.996147527, 1916524903.227499962]: Starting TiltAntenna: 0.000000 degrees (0.000000 radians)
[ INFO] [1667849277.996464409, 1916524903.227499962]: Starting PanAntenna: 0.000000 degrees (0.000000 radians)
[ INFO] [1667849277.996529644, 1916524903.227499962]: PLEXIL: TEST: 1
[ INFO] [1667849277.996578206, 1916524903.227499962]: Starting PanAntenna: 183.346000 degrees (3.199991 radians)
[ERROR] [1667849291.037170298, 1916524933.329999924]: PanAntenna timed out
[ERROR] [1667849291.037211654, 1916524933.329999924]: PanAntenna failed. Ended at -176.654000 degrees, goal was 183.346000.
[ERROR] [1667849291.037389314, 1916524933.329999924]: PLEXIL: Pan: FAILED Goal: 183.346 Actual: -176.6539998302
[ INFO] [1667849291.037446712, 1916524933.329999924]: PLEXIL: TEST: 1
[ INFO] [1667849291.037526816, 1916524933.329999924]: Starting PanAntenna: -183.346000 degrees (-3.199991 radians)
[ INFO] [1667849291.238012308, 1916524933.930000067]: PLEXIL: Pan: SUCCEEDED Goal: -183.346 Actual: -180.702395206843
[ INFO] [1667849291.238109642, 1916524933.930000067]: PLEXIL: TEST: 2
[ INFO] [1667849291.238190060, 1916524933.930000067]: Starting PanAntenna: 0.000000 degrees (0.000000 radians)
[ERROR] [1667849301.355856473, 1916524964.032500029]: PanAntenna timed out
[ERROR] [1667849301.355901110, 1916524964.032500029]: PanAntenna failed. Ended at -360.000000 degrees, goal was 0.000000.
[ERROR] [1667849301.355969100, 1916524964.032500029]: PLEXIL: Pan: FAILED Goal: 0 Actual: -359.99999997
[ INFO] [1667849301.356053688, 1916524964.032500029]: PLEXIL: TEST: 2
[ INFO] [1667849301.356128531, 1916524964.032500029]: Starting PanAntenna: -45.000000 degrees (-0.785398 radians)
[ERROR] [1667849311.474545450, 1916524994.130000114]: PanAntenna timed out
[ERROR] [1667849311.474583339, 1916524994.130000114]: PanAntenna failed. Ended at -404.999998 degrees, goal was -45.000000.
[ERROR] [1667849311.474677280, 1916524994.130000114]: PLEXIL: Pan: FAILED Goal: -45 Actual: -404.999997507995
[ INFO] [1667849311.474733846, 1916524994.130000114]: PLEXIL: TEST: 2
[ INFO] [1667849311.474808787, 1916524994.130000114]: Starting PanAntenna: 10.000000 degrees (0.174533 radians)
[ERROR] [1667849321.599128352, 1916525024.230000019]: PanAntenna timed out
[ERROR] [1667849321.599166315, 1916525024.230000019]: PanAntenna failed. Ended at -350.000001 degrees, goal was 10.000000.
[ERROR] [1667849321.599239540, 1916525024.230000019]: PLEXIL: Pan: FAILED Goal: 10 Actual: -350.000000578252
[ INFO] [1667849321.599304119, 1916525024.230000019]: PLEXIL: TEST: 2
[ INFO] [1667849321.599386997, 1916525024.230000019]: Starting PanAntenna: 0.000000 degrees (0.000000 radians)
[ERROR] [1667849331.669406749, 1916525054.329999924]: PanAntenna timed out
[ERROR] [1667849331.669441033, 1916525054.329999924]: PanAntenna failed. Ended at -360.000000 degrees, goal was 0.000000.
[ERROR] [1667849331.669505272, 1916525054.329999924]: PLEXIL: Pan: FAILED Goal: 0 Actual: -359.999999967697
[ INFO] [1667849331.669577193, 1916525054.329999924]: PLEXIL: TEST: 2
[ INFO] [1667849331.669668381, 1916525054.329999924]: Starting TiltAntenna: 10.000000 degrees (0.174533 radians)
[ INFO] [1667849331.904108142, 1916525055.032500029]: PLEXIL: Tilt: SUCCEEDED Goal: 10 Actual: 7.92322101076209
[ INFO] [1667849331.904169348, 1916525055.032500029]: PLEXIL: TEST: 3
[ INFO] [1667849331.904234573, 1916525055.032500029]: Starting TiltAntenna: 20.000000 degrees (0.349066 radians)
[ INFO] [1667849332.171570163, 1916525055.829999924]: PLEXIL: Tilt: SUCCEEDED Goal: 20 Actual: 17.5756048731797
[ INFO] [1667849332.171645505, 1916525055.829999924]: PLEXIL: TEST: 4
[ INFO] [1667849332.171717487, 1916525055.829999924]: Starting TiltAntenna: -10.000000 degrees (-0.174533 radians)
[ INFO] [1667849332.839281069, 1916525057.829999924]: PLEXIL: Tilt: SUCCEEDED Goal: -10 Actual: -7.26288646047045
[ INFO] [1667849332.839339036, 1916525057.829999924]: PLEXIL: TEST: 5
[ INFO] [1667849332.839403784, 1916525057.829999924]: Starting TiltAntenna: 89.380000 degrees (1.559975 radians)
[ INFO] [1667849335.319370556, 1916525065.230000019]: PLEXIL: Tilt: SUCCEEDED Goal: 89.38 Actual: 86.9534065679976
[ INFO] [1667849335.319464873, 1916525065.230000019]: PLEXIL: TEST: 6
[ INFO] [1667849335.319565790, 1916525065.230000019]: Starting TiltAntenna: -89.380000 degrees (-1.559975 radians)
[ INFO] [1667849340.018505314, 1916525079.130000114]: PLEXIL: Tilt: SUCCEEDED Goal: -89.38 Actual: -87.4168061894679
[ INFO] [1667849340.018576427, 1916525079.130000114]: PLEXIL: TEST: 7
[ INFO] [1667849340.018656634, 1916525079.130000114]: Starting TiltAntenna: 45.000000 degrees (0.785398 radians)
[ INFO] [1667849343.521198074, 1916525089.529999971]: PLEXIL: Tilt: SUCCEEDED Goal: 45 Actual: 43.1803835846703
[ INFO] [1667849343.521303210, 1916525089.532500029]: PLEXIL: TEST: 8
[ INFO] [1667849343.521395432, 1916525089.532500029]: Starting TiltAntenna: 0.000000 degrees (0.000000 radians)
[ INFO] [1667849344.626296247, 1916525092.827500105]: PLEXIL: Tilt: SUCCEEDED Goal: 0 Actual: 2.67030282123032
[ERROR] [1667849344.626381921, 1916525092.827500105]: PLEXIL: TestAntennaMovement: FAILED, 8/13 Tests Succeeded
```

kmdalal commented:
Two separate bugs have been fixed, and the test should now pass.

AstroStucky commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -308,6 +303,14 @@ void OwInterface::antennaFaultCallback
+  static double pi = R2D * M_PI;
+  static double tau = pi * 2.0;
+
+  return fmod(angle + pi, tau) - pi;
```
TIL C++ `fmod` != Python `%` (floating-point modulus)

When the numerator is negative, Python's floating-point modulus operator will use a negative quotient, so that the remainder that is returned is always positive. However, in the case of cmath's floating-point modulus function, the quotient is always selected to be positive, which means the remainder returned by the function will be negative so long as the denominator is positive. The result is that this function can return angles as low as -360. I think this is where the bug came from. This can be accounted for with the addition of an if-statement to handle the negative case. 
**Source:** https://stackoverflow.com/questions/11498169/dealing-with-angle-wrap-in-c-code
```suggestion
  double x = fmod(angle + pi, tau);
  if (x < 0)
    x += tau;
  return x - pi;
```

kmdalal replied:
Thanks, done.  This [StackOverflow article](https://stackoverflow.com/questions/11498169/dealing-with-angle-wrap-in-c-code) helped verify.

kmdalal commented:
Removed unused import, and added commented-out code for debugging an upcoming Jira.

AstroStucky commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -401,6 +385,12 @@ void OwInterface::managePanTilt (const string& opname,
+
+bool OwInterface::anglesEquivalent (double deg1, double deg2, double tolerance)
+{
+  return normalize_degrees(fabs(deg1 - deg2)) <= tolerance;
```
Now that normalize_degrees is fixed, this should be changed to the form that takes absolute valued after normalizing.

kmdalal replied:
Okay, done, and the tests still pass.

AstroStucky commented on ow_plexil/src/plans/TestAntennaMovement.plp:
```diff
@@ -6,38 +6,40 @@
   }
   else {
-    log_error("TestAntennaMovement: FAILED, ", TestCount, "/13 Tests Succeeded");
+    log_error("TestAntennaMovement: FAILED, ", Wins, "/13 Tests Succeeded");
```
[optional] Could the literal 13 just be Count?

kmdalal replied:
Good catch.  I got rid of the literal everywhere, and had to adjust the count logic a bit for things to work.

AstroStucky reviewed (approved):
Looks good!
